from flask import Flask, request, jsonify
from flask_sqlalchemy import SQLAlchemy
from flask_cors import CORS

app = Flask(__name__)
CORS(app)
app.config['SQLALCHEMY_DATABASE_URI'] = 'mysql+mysqlconnector://root:root@localhost:3306/users'
app.config['SQLALCHEMY_TRACK_MODIFICATIONS'] = False

db = SQLAlchemy(app)

class User(db.Model):
    __tablename__ = 'users'
    username = db.Column(db.String(255), primary_key=True)
    email = db.Column(db.String(255), nullable=False)
    favourites = db.Column(db.String(255))

    def __init__(self, username, email, favourites=''):
        self.username = username
        self.email = email
        self.favourites = favourites

@app.before_first_request
def create_tables():
    db.create_all()

@app.route('/insert_users', methods=['POST'])
def insert_users():
    data = request.get_json()
    try:
        for item in data:
            username = item['username']
            email = item['email']
            favourite = item.get('favourite', '')  # Use .get to avoid KeyError if 'favourite' is missing
            
            user = User.query.filter_by(username=username).first()  # Filter by username only, assuming it's unique
            if user:
                # Check and append favourite if not already in user.favourites
                if favourite and favourite not in user.favourites.split(','):
                    user.favourites += ',' + favourite if user.favourites else favourite
            else:
                new_user = User(username=username, email=email, favourites=favourite)
                db.session.add(new_user)
            db.session.commit()
        return jsonify({'message': 'Users inserted successfully'}), 201
    except Exception as e:
        db.session.rollback()
        return jsonify({'error': str(e)}), 500

if __name__ == '__main__':
    app.run(port=5002, debug=True)